### PR TITLE
Fix m3u list parsing with /r/n end of line

### PR DESCRIPTION
**Pull request by daschuer**
This fixes a regression form
https://github.com/mixxxdj/mixxx/commit/a006552fec6877facc4aaa22080cc67d6699941d

### CODE REVIEW

**Comment by Swiftb0y** [comment]
Wdyt about adding a test to make sure this regression does not happen again? Should be relatively straightforward. 
I guess only `main` is affected right?

File: src/library/parserm3u.cpp
Comment (Holzhaus) [inline src/library/parserm3u.cpp:83]: Deprecated in Qt6, the docs recommend to use QRegularExpression instead: https://doc.qt.io/qt-6/qregexp.html#details

**Review by Holzhaus — CHANGES_REQUESTED** [review]

File: src/library/parserm3u.cpp
Comment (uklotzde) [reply src/library/parserm3u.cpp:83]: Should also be extracted into a constant and compiled only once if the matching is repeated on different inputs.

Comment (daschuer) [reply src/library/parserm3u.cpp]: Done.

**Comment by daschuer** [comment]
Test is in place now.

However I have discovered a kind of regression due to: 

https://github.com/mixxxdj/mixxx/commit/c1dcdccd3241475630b26503ebb079620338d248

@uklotzde you have moved the existence check from a central point into the m3u parser. This removed the existence check from all other playlist types. What is you recommendation to fix this issue?

**Comment by uklotzde** [comment]
> Test is in place now.
> 
> However I have discovered a kind of regression due to:
> 
> [c1dcdcc](https://github.com/mixxxdj/mixxx/commit/c1dcdccd3241475630b26503ebb079620338d248)
> 
> @uklotzde you have moved the existence check from a central point into the m3u parser. This removed the existence check from all other playlist types. What is you recommendation to fix this issue?

I don't spot any regression? Invoking TrackFile::checkExists() when needed instead of moving an anonymous boolean value around is safer and less error prone.

**Comment by daschuer** [comment]
I just had a closer look. The issue is that ParserCsv::parse() has now no existence check. Not existing tracks are leading to a debug assertion violation:
```
[91mcritical[0m [[97mMain[0m] DEBUG ASSERT: "sourceSynchronizedAt.isValid()" in function bool mixxx::TrackRecord::replaceMetadataFromSource(mixxx::TrackMetadata&&, const QDateTime&) at /home/sperry/workspace/2.3/src/track/trackrecord.cpp:184
```
I have misread the code around ParserPls::parse() which also removes not existing tracks. Now I have found the check. 

The anonymous boolean was only a temporary solution, to show the issue. 

I think this requires some more refactoring to make the code testable and keep the code in a good shape at the same time. A creeping scope of this PR :-/ .. but OK, right?

File: src/test/playlisttest.cpp
Comment (Swiftb0y) [inline src/test/playlisttest.cpp:84]: the size check after the EXPECT_EQ seems unnecessary.

Comment (daschuer) [reply src/test/playlisttest.cpp]: Done.

**Comment by daschuer** [comment]
This fixes now also the debug assert with non existing csv entries. 
I have also added tests for pls and csv. 
csv files can now also be used if no or a translated header line is found.

File: src/library/parser.cpp
Comment (Swiftb0y) [inline src/library/parser.cpp:34]:
```suggestion
    for (const auto& location : qAsConst(locations)) {
```
`qAsConst` if you disagree that making the QList const to begin with is a good idea.

File: src/test/playlisttest.cpp
Comment (Swiftb0y) [reply src/test/playlisttest.cpp]: Thanks

**Comment by daschuer** [comment]
I have now introduced the named function: parseAllLocations() this is even better than the anonymous lambda.

**Comment by Swiftb0y** [comment]
well, the lambda was basically the C++ workaround for what would otherwise just be a simple block expression in rust, but extracting it into a function is good too.

File: src/library/parserpls.cpp
Comment (Swiftb0y) [inline src/library/parserpls.cpp:13]: why the ptr instead of a const-ref?

File: src/test/playlisttest.cpp
Comment (Swiftb0y) [inline src/test/playlisttest.cpp:72]:
```suggestion
    const QList<QString> entries = ParserM3u().parseAllLocations(m3uFile.fileName());
```

File: src/test/playlisttest.cpp
Comment (Swiftb0y) [inline src/test/playlisttest.cpp:78]: generally favor `QList.at()` over `operator[]` iirc

**Review by Swiftb0y — COMMENTED** [review]
This PR seems done, right?
@Holzhaus we still need your approval/dismissal since you requested changes.

File: src/test/playlisttest.cpp
Comment (Holzhaus) [reply src/test/playlisttest.cpp:78]: Why? If the test fails, let it crash.

File: src/library/parserpls.cpp
Comment (Holzhaus) [reply src/library/parserpls.cpp:13]: It modifies the stream (moves the seek position) so I think a ptr is semantically appropriate.

**Comment by Holzhaus** [comment]
> This PR seems done, right?
> @Holzhaus we still need your approval/dismissal since you requested changes.

Yes, I just re-reviewed and chose comment instead of request changes to clarify I'm okay with merging.

File: src/test/playlisttest.cpp
Comment (Swiftb0y) [reply src/test/playlisttest.cpp:78]: .at() does not perform bounds checking. It's rather preferred because it is const and thus guarantees that the implicitly shared container will not be detached (which does not matter in this particular context but still good practice I'd argue).

File: src/library/parserpls.cpp
Comment (Swiftb0y) [reply src/library/parserpls.cpp:13]: Ah, right, so the text stream must be mutable, but that still does not explain why were passing a pointer instead of a (mutable) reference. AFAIK they are semantically the same, so is the pointer just used because it was there in the first place?

Comment (Swiftb0y) [reply src/library/parserpls.cpp:13]: Keep in mind that this is not a request for changes, I'm must asking because I'd like improve my C++ knowledge.

File: src/test/playlisttest.cpp
Comment (Swiftb0y) [reply src/test/playlisttest.cpp:72]: I really wish there was a compiler warning for improving const correctness, but according the only stackoverflow answer I found, that seems do be deemed useless by the majority of C++ devs...

**Comment by Swiftb0y** [comment]
> > This PR seems done, right?
> > @Holzhaus we still need your approval/dismissal since you requested changes.
> 
> Yes, I just re-reviewed and chose comment instead of request changes to clarify I'm okay with merging.

Thanks for clarification. For some reason, GitHub does not show that when you look into the reviewers section at the top right of the PR. Maybe you need to dismiss the review or overwrite it via an explicit approval?

File: src/library/parser.cpp
Comment (Swiftb0y) [inline src/library/parser.cpp:21]: consider adding ``[[nodiscard]]`` to all of the new pure functions

File: src/library/parserpls.cpp
Comment (daschuer) [reply src/library/parserpls.cpp:13]: I will change "stream" to "pStream" ...

From the C++ rules we may pass here a non const reference: 
`QString getFilePath(QTextStream& stream);` But it is a matter of code style.  

Than we see: 
`stream.readLine();`
and not 
`pStream->readLine();`

The later makes clear, that we do not have a local copy and all work with the pointed object will have an outside effect. 

In case of `QString getFilePath(cosnt QTextStream& stream);` it does not matter if it is a copy by value or a reference, It has never an outside effect, because of the constantnes 
  
This a fundamental difference between C++ and Java or C#, where objects are always a references and where thy are no pointers. 

This was also part of the Google style. However they have recently dropped the strict rule.

See also: 
https://github.com/mixxxdj/mixxx/wiki/Coding-Guidelines#non-const-references

Comment (Swiftb0y) [reply src/library/parserpls.cpp:13]: Thanks for the explanation. Stylewise, when passing a pointer, while passing `nullptr` would not make sense, it is still permitted. Wouldn't it make sense to use a mutable reference instead since those can't be null?
Moreover, in this particular case, we don't treat the incoming QTextStream as an output variable, but rather consume the textstream and expect it to not be used in the caller afterwards, correct? Because then the parameter rather seems like it should semantically be a `QTextStream&&`

File: src/test/playlisttest.cpp
Comment (daschuer) [reply src/test/playlisttest.cpp:78]: There is literally no difference: 

```
inline const T &QList<T>::operator[](int i) const
{ Q_ASSERT_X(i >= 0 && i < p.size(), "QList<T>::operator[]", "index out of range");
 return reinterpret_cast<Node *>(p.at(i))->t(); }
```
```
inline const T &QList<T>::at(int i) const
{ Q_ASSERT_X(i >= 0 && i < p.size(), "QList<T>::at", "index out of range");
 return reinterpret_cast<Node *>(p.at(i))->t(); }
```

With [] we just to be careful not to use the non const variant which is not obvious. From this point of view preferring at() is a good Idea. It is also recommended by Qt: 
> If you do not want to modify the list you should use QList::at().

File: src/library/parser.cpp
Comment (daschuer) [reply src/library/parser.cpp:21]: We consider [[nodiscard]] for pure functions as unnecessary verbosity. It is not the business of the function to restrict its usage. 

See: https://github.com/mixxxdj/mixxx/wiki/Coding-Guidelines#attribute-nodiscard

I think the usage in the standard is a good patter see: 
https://en.cppreference.com/w/cpp/language/attributes/nodiscard

It is used for empty() where it should have been named isEmpty() to distinguished it from clear() and for functions that forwards the ownership like new() and discarding the return value will lead to a memory leak.

**Comment by daschuer** [comment]
Done.

File: src/library/parser.cpp
Comment (Swiftb0y) [reply src/library/parser.cpp:21]: If you think so...

File: src/test/playlisttest.cpp
Comment (Swiftb0y) [reply src/test/playlisttest.cpp]: Right, I already said that it does not matter here, but we should still promote good coding standards, even in our test code which is why I brought it up.

File: src/library/parserpls.cpp
Comment (daschuer) [reply src/library/parserpls.cpp]: > Wouldn't it make sense to use a mutable reference instead since those can't be null?

That is a valid argument for the other code style. Both styles are valid, we prefere the reliable value vs. pointer syntax style. By the way: in C++ a mutable reference can be null. It is only by convention never null. Pointers and references are exactly the same if you look at the produced assembler code.

Comment (Swiftb0y) [reply src/library/parserpls.cpp]: > That is a valid argument for the other code style. Both styles are valid, we prefere the reliable value vs. pointer syntax style.

Thanks for the clarification. I guess I'd prefer the the other style but I can adopt the mixxx style as well. 

>  By the way: in C++ a mutable reference can be null. It is only by convention never null. Pointers and references are exactly the same if you look at the produced assembler code.

I know, but I thought the compiler does not allow references to be set to null (at compile time, there aren't any runtime checks unless you use something like `gsl::not_null`).

**Review by Holzhaus — APPROVED** [review]

File: src/library/parsercsv.h
Comment (daschuer) [reply src/library/parsercsv.h:23]: It defaults to private. 
But to be honest, I have just forget to write it explicitly. I will fix it.

Comment (Swiftb0y) [reply src/library/parsercsv.h:23]: Thanks, that was my concern. Seems like no outside component depends on the inheritance but I think it makes sense for the relationship to be public regardless.

Comment (daschuer) [reply src/library/parsercsv.h]: Done.

**Comment by Swiftb0y** [comment]
Any idea whats tripping up clang-format? Related to #4557?

**Comment by uklotzde** [comment]
> Any idea whats tripping up clang-format? Related to #4557?

We now require `Standard: c++17` and `>>` is valid syntax since C++11. Now legacy code using `> >` gets fixed.

**Comment by uklotzde** [comment]
It is probably a good idea to replace all occurrences of `> >` with `>>` in a subsequent PR. It is a small and trivial change that does not need to be deferred by the *no mass reformatting* rule.

**Comment by daschuer** [comment]
> It is probably a good idea to replace all occurrences of > > with >> in a subsequent PR.

I have no interest to do that. The `> >` does not hurt my eyes and there is a good chance of creating conflicts with existing PRs.

File: src/test/playlisttest.cpp
Comment (Holzhaus) [inline src/test/playlisttest.cpp:61]: Everywhere else we use PascalCase for the test names, let's keep this pattern here:

```suggestion
TEST_F(PlaylistTest, M3UEndOfLine) {
```

**Comment by Holzhaus** [comment]
> > It is probably a good idea to replace all occurrences of > > with >> in a subsequent PR.
> 
> I have no interest to do that.

No problem, I took care of it here: https://github.com/mixxxdj/mixxx/pull/4559

**Comment by daschuer** [comment]
Conflict resolved, merge?

**Pull request merged**
